docs/contributor_guide/architecture.md

```diff
@@ -0,0 +1,119 @@
+# Architecture overview
+
+JupyterGIS is a JupyterLab extension (based on the structure defined by
+[jupyterlab/extensions-cookiecutter-ts](https://github.com/jupyterlab/extension-cookiecutter-ts).
+
+Its architecture is based on QuantStack's
+[JupyterCAD](https://github.com/QuantStack/JupyterCAD) architecture.
+
+
+## JupyterLab
+
+### Backbone
+
+...
+
+
+## JupyterGIS components
+
+### Directory structure
+
+JupyterGIS is a monorepo containing TypeScript and Python packages.
+
+* `python/`: The Python packages! These may include some TypeScript as well (_TODO: why?
+  Link to JupyterLab extension doc?_).
+  * `jupytergis`: A metapackage including `jupytergis_core`, `jupytergis_lab`,
+    `jupytergis_qgis`, `jupyter-collaboration`, and `jupyterlab`.
+  * `jupytergis_core`: Python methods specific to JupyterLab (not JupyterLite).
+    * :star: `src/jgisplugin/plugins.ts`: Where all the commands we've defined are
+      registered in TypeScript land -- defines behaviors for menu items.
+    * `src/jgisplugin/plugins.ts`: Where commands are linked to layer and source context
+      menus.
+  * `jupytergis_lab`: _TODO: Ask Martin_.
+    * :star: `jupytergis_lab/notebook/gis_document.ts`: Entrypoint for Python API.
+    * `src/notebookrenderer.ts`: Enables rendering JupyterGIS in a Notebook.
+  * `jupytergis_lite`: A metapackage for deployment and testing of JupyterGIS in
+    JupyterLite.
+  * `jupytergis_qgis`: Enables importing and exporting QGIS project files.
+* `packages/`: The TypeScript packages.
+  * `base/`:
+    * `src/index.ts`: Entrypoint for Typescript code.
+    * :star: `src/constants.ts`: Defines all the JupyterGIS command names and
+      corresponding icons.
+    * :star: `src/commands.ts`: For each command, defines the labels, the behaviors, and
+      in what contexts each command is available.
+    * :star: `src/mainview/index.ts`: Entrypoint for the main map view.
+    * `src/toolbar/widget.tsx`: Where commands are linked to the toolbar.
+    * `src/formbuilder/`: Base React components for forms.
+  * :star: `schema/`: The source of truth for data structures in JupyterGIS. Python
+    classes and Typescript types are generated from this.
```

Comment:
This documentation feels like it should be auto-generated from metadata files in the source code directories. For example, from front-matter in READMEs. @batpad @arjxn-py do you know of anything that can do that?

Reply:
:+1: to have the documentation in the READMEs.
[`myst_parser`](https://myst-parser.readthedocs.io/en/latest/) should be able to handle.

Reply:
I was hoping for a way to generate a directory structure like the one @batpad and @arjxn-py and I wrote above from the actual directory structure. I don't see a way to do that from the myst_parser docs or sphin docs, do you know what I might be missing?

Reply:
Sorry, I misunderstood the question.
Maybe we should use an API doc generator, like [typedoc](https://typedoc.org/api/index.html).

Reply:
Looks like they ruled that out: https://github.com/TypeStrong/typedoc/issues/1659

Maybe I can write an autodoc extension that does this? :thinking:

Reply:
Took a stab at this here, forgot to share :) https://github.com/SchmidtDSE/autodoc-directory-structure